### ISSUE TITLE
[CI] Restartable pipeline

### ISSUE THREAD

**Issue by st-pasha** [h2oai/datatable#2267]
Currently, the Jenkins pipeline is organized as a sequence of steps such that whenever one of the steps fail, the pipeline is terminated, and then has to be restarted from the very beginning once the issue is addressed. This is problematic, especially if the steps before the failure have been successful, and re-running them will waste significant amount of time.

This proposal suggests to re-organize the pipeline around the idea that each step in the pipeline should persist its result in S3, and then if restarted, Jenkins should skip those stages that have been successful before.

The suggested structure is the following:
1. First, the master task performs git checkout, and immediately builds an `sdist` of the datatable. Building `sdist` requires only standard python (3.5+), and can be done either in a docker, or in a virtual env, or without any virtualization at all. The `sdist` produces a single `.tar.gz` file, which contains all files that are considered "source" of the datatable. This sdist is stored in S3.
2. Similarly, we tarball all tests in the datatable repo, and also stash them in S3.
3. The build stage(s) produce binary wheels of the datatable based on the `sdist` file. The output wheels are stored in S3, tagged with the sha-hash of the sdist. The stage can be skipped if the wheels already exist.
4. Each test stage grabs the wheel file, and the tests-tarball from S3; runs the tests; and stores the "success" marker on S3 (which could be a 0-size file). The stage can be skipped if the corresponding marker (tagged with sha of the wheel and the tests-tarball) already exists in S3.
5. The final stage verifies the existence of all wheels/test results in S3, and performs any epilogue actions.

This structure ensures that: 
- if the source code changes, the entire pipeline is rebuilt (same as now);
- if only the tests change, then all tests are re-run but wheels are not rebuilt;
- if only auxiliary files change (docs, jenkins scripts, git support files, etc), or no files change at all (e.g. Jenkins job restart), then only steps that failed before need to rerun.

**Comment by mmalohlava** [comment]
@surenH2oai this can be good starting task

**Comment by surenH2oai** [comment]
Yes.  let me plan this.  When the Polish team is in MTNV lets sync with @st-pasha.

**Comment by surenH2oai** [comment]
Assigning to Krakow team.

**Comment by surenH2oai** [comment]
we need to design the solution and usage in to ensure just restarting the pipeline from last failure after a fix may leak regression if fully not rerun.  I guess this is a developer choice based on their experience of the code change. So we should be fine.

**Comment by st-pasha** [comment]
@surenH2oai The proposal is already regression-proof: any change to code within a PR would trigger a full rerun of the pipeline.

**Comment by st-pasha** [comment]
To be more specific, this is an approximate outline of the pipeline:

1. Git checkout PR branch -> dir with datatable sources -> make sdist + stests + sdocs [+ ...]. Each of these tasks creates a `.tar.gz` file, and we'll refer to sha256 hashes of those files as "sdist-sha", "stests-sha", etc. 
2. File `sdist.tar.gz` (it will actually be called `datatable.{version}.tar.gz`) is stored in S3 under the name `s3://datatable/tmp/{sdist-sha}/sdist.tar.gz`, unless it already exists there.
3. File `stests.tar.gz` containing all files from `tests/` directory is stored in S3 under the name `s3://datatable/tmp/{sdist-sha}/{stests-sha}/stests.tar.gz` (unless the file already exists); and similarly for `sdocs`. In the future we may want to split tests into several disjoint groups such that changes to one group wouldn't affect the need to test the other.
4. **Build stage**: for each platform/python version that we need to build, check for the existence of the file `s3://datatable/tmp/{sdist-sha}/datatable.{pyabi}-{platform}.whl`, and if it doesn't exist -- build the corresponding wheel from `sdist.tar.gz` and store the wheel in S3.
5. **Tests stage**: for each test stage, first check whether the marker file `s3://datatable/tmp/{sdist-sha}/{stests-sha}/tests-{stage}.done` exists. If it doesn't, then retrieve the necessary wheel file and the stests tarball, run the tests, and in case of success store the aforementioned marker file into the S3.
6. **Publish stage**: if all tests ran successfully, copy all wheels and the sdist into `s3://datatable/builds/{version}/`. This is only needed for master builds.

Note that the namespace depends only on `{sdist-sha}`, and not on git-branch, nor on build number. We want to allow separate PRs to share the build/test results if those PRs do not touch the source code (such as documentation or configuration changes). Also, a "master" build after a successful PR should be able to skip most of the stages.

I haven't mentioned much about the `sdocs` tarball, because currently we don't have any doc tests, and don't even try to build the docs automatically. However, this situation should be fixed going forward, and there must be some sanity checks for the documentation too.

**Comment by st-pasha** [comment]
Note: the aforementioned procedure must be updated to account for the fact that both sdist and wheels contain a volatile `_build_info.py` file, whose content changes upon every build. This file should be excluded when computing "source sha", but included in actual sdist and wheels.

Thus, we'll have to produce separately a tarball of source files, and the "sdist" file which is a function of the source files, and the wheel files which are based on the sdist.

The release builds should be separate from the "dev" builds, even if their sources have same SHAs.
